### PR TITLE
Test cleanups/robustifications

### LINKED ISSUE #13717: @nondestructive test cleanups
This is a meta-bug for things that cross my mind (and my eyes in PR reviews):

 - [x] Introduce a helper for sedding a file and restoring it: PR #13725 
 - [x] Introduce a helper for backing up/restoring a directory:  PR #13751 
 - [ ] Generalize `TestMachines.addIface()` (introduced in PR #13699): PR #13788
 - [x] Fix SELinux disabling in `installWithErrorTest` (use `chcon`): We may just be able to [drop this completely](https://github.com/cockpit-project/cockpit/pull/13699#discussion_r392837894) ← we did that
 - [x] Try and drop allowBugErrors() in machineslib.py: PR #13747

### PR DESCRIPTION
**Pull request by martinpitt**
Fixes some parts of #13717

### CODE REVIEW

**Comment by martinpitt** [comment]
Ah yes, I need to make that helper [cope with nonexisting files](https://logs.cockpit-project.org/logs/pull-13725-20200316-132356-959a07d0-fedora-31/log.html#2-2)

**Comment by martinpitt** [comment]
Just landed #13699, I should update this PR again to include the backups/restores from that PR.

**Comment by martinpitt** [comment]
/home is a symlink on CoreOS, so [restore_dir() needs adjustment](https://logs.cockpit-project.org/logs/pull-13725-20200316-182045-8a47a203-fedora-coreos/log.html). Fixed.

**Comment by martinpitt** [comment]
Sometimes the [machines cleanup is falling over badly](https://logs.cockpit-project.org/logs/pull-13725-20200317-064132-b3ffcfc9-fedora-32/log.html#216-2), it seems that the `vmNotInstalled` domain is still living in libvirtd's brain. I can't reproduce this locally, and it's not failing reliably (e. g. in the previous round it happened on ubuntu-2004). I pushed another commit which may help with this, let's see how this fares.

**Comment by martinpitt** [comment]
Lots of [AppArmor errors on debian-testing](https://logs.cockpit-project.org/logs/pull-13725-20200317-085237-d3b351ea-debian-testing/log.html#229), I didn't see them before. Weird that it doesn't like /var/tmp/cockpittest/ but it did like /mnt/libvirt !? I'll take a closer look.

**Comment by marusak** [comment]
> Weird that it doesn't like /var/tmp/cockpittest/ but it did like /mnt/libvirt !?

Heh yeah, I also firstly used `/var/tmp` but tests were not happy, so I just used `/mnt`. Some storage related things (like pools or volumes?) were failing. I however did not investigate it.

**Comment by martinpitt** [comment]
Hm, that wasn't it -- with /mnt I get the same locally:

audit: type=1400 audit(1584443778.236:61): apparmor="DENIED" operation="open" profile="libvirt-c821ed97-8f5c-4938-b437-b62f0887ea6b" name="/mnt/cockpittest/vm_one/mydiskofpoolone_temporary" pid=8208 comm="qemu-system-x86" requested_mask="r" denied_mask="r" fsuid=64055 ouid=64055

There's also nothing about /mnt in the AppArmor policy that would apply to libvirt. Curiously, the only message that it complains about locally is `invalid or unusable locale: en_US.UTF-8`, but that doesn't even happen in CI. So this will need a closer look, just moving to /mnt is not the fix.

**Comment by martinpitt** [comment]
I removed the more intrusive commits, landing the simpler ones first. Let's tread more slowly then.

**Comment by marusak** [comment]
I suggest picking up commit `test: Clean up inheritance of TestMachines` into separate PR.
Together with https://github.com/cockpit-project/cockpit/pull/13740 it fixes that annoying `mkdir /mnt/libvirt failed: directory exists`. See [here](https://github.com/cockpit-project/cockpit/pull/13733#issuecomment-600251086). That is high priority

**Comment by martinpitt** [comment]
Ack, split out the inheritance fix to #13741

**Comment by martinpitt** [comment]
I took out the inheritance fix here, as it now has its own separate PR.

[This](https://logs.cockpit-project.org/logs/pull-13725-20200317-190858-df81c5ec-debian-stable/log.html#70-1) is an interesting failure -- we do a lot of error noise on the Machines page. I'll send a separate fix for that.

File: test/common/testlib.py [hunk 1]
```diff
@@ -1192,6 +1192,23 @@ def settle_cpu(self):
...
+        This is safe for @nondestructive tests, the file will be restored during cleanup.
+
+        The optional apply_change_action will be run both after sedding and after restoring the file.
+        '''
+        m = self.machine
+        m.execute("sed -i.cockpittest '{0}' {1}".format(expr, path))
+        if apply_change_action:
+            m.execute(apply_change_action)
+
+        if self.is_nondestructive():
```
Comment (marusak) [inline test/common/testlib.py]: Do we actually need to check if is non destructive? Cleaning up always cannot do any harm, even when we destroy the VM right after. But fair enough.

Comment (martinpitt) [reply test/common/testlib.py]: It's not doing harm, but it is wasting an extra round of ssh and running IO intense commands, I'd rather avoid these if right after that we'll kill the VM anyway.

**Comment by martinpitt** [comment]
Some [virsh default network trouble](https://logs.cockpit-project.org/logs/pull-13725-20200318-101739-9bb0a27e-ubuntu-1804/log.html#72-2) on ubuntu-1804. Let's retry this, to make sure that the first commit didn't subtly break this.

**Comment by martinpitt** [comment]
I am now really [annoyed](https://logs.cockpit-project.org/logs/pull-13725-20200318-101739-9bb0a27e-debian-stable/log.html#243) [by these](https://logs.cockpit-project.org/logs/pull-13725-20200318-101739-9bb0a27e-rhel-8-2-distropkg/log.html#243-2) [timeouts](https://logs.cockpit-project.org/logs/pull-13725-20200318-101739-9bb0a27e-ubuntu-stable/log.html#244-2). It makes absolutely no sense to me why making the waiting for the default network much more expensive would magically fix it.. I'll take out that bit, it's not that important.

**Comment by martinpitt** [comment]
OK, so that previous wait simplification [was not it](https://logs.cockpit-project.org/logs/pull-13725-20200318-124232-2b8fedd9-ubuntu-1804/log.html#101). I'm going to split out some bits.

**Comment by martinpitt** [comment]
OK, that's much better. Apparently the more aggressive cleanup is breaking something.

**Pull request merged**